zio-http/shared/src/main/scala/zio/http/Server.scala

```diff
@@ -200,22 +200,25 @@ object Server extends ServerPlatformSpecific {
   object Config {
     lazy val config: zio.Config[Config] = {
       SSLConfig.config.optional ++
-        zio.Config.string("binding-host").optional ++
-        zio.Config.int("binding-port").withDefault(Config.default.address.getPort) ++
-        zio.Config.boolean("accept-continue").withDefault(Config.default.acceptContinue) ++
-        zio.Config.boolean("keep-alive").withDefault(Config.default.keepAlive) ++
-        Decompression.config.nested("request-decompression").withDefault(Config.default.requestDecompression) ++
-        ResponseCompressionConfig.config.nested("response-compression").optional ++
-        RequestStreaming.config.nested("request-streaming").withDefault(Config.default.requestStreaming) ++
-        zio.Config.int("max-initial-line-length").withDefault(Config.default.maxInitialLineLength) ++
-        zio.Config.int("max-header-size").withDefault(Config.default.maxHeaderSize) ++
-        zio.Config.boolean("log-warning-on-fatal-error").withDefault(Config.default.logWarningOnFatalError) ++
-        zio.Config.duration("graceful-shutdown-timeout").withDefault(Config.default.gracefulShutdownTimeout) ++
-        zio.Config.duration("idle-timeout").optional.withDefault(Config.default.idleTimeout) ++
-        zio.Config.boolean("avoid-context-switching").withDefault(Config.default.avoidContextSwitching) ++
-        zio.Config.int("so-backlog").withDefault(Config.default.soBacklog) ++
-        zio.Config.boolean("tcp-nodelay").withDefault(Config.default.tcpNoDelay)
-
+        zio.Config.string.nested("binding", "host").optional ++
+        zio.Config.int.nested("binding", "port").withDefault(Config.default.address.getPort) ++
+        zio.Config.boolean.nested("accept", "continue").withDefault(Config.default.acceptContinue) ++
+        zio.Config.boolean.nested("keep", "alive").withDefault(Config.default.keepAlive) ++
+        Decompression.config.nested("request", "decompression").withDefault(Config.default.requestDecompression) ++
+        ResponseCompressionConfig.config.nested("response", "compression").optional ++
+        RequestStreaming.config.nested("request", "streaming").withDefault(Config.default.requestStreaming) ++
+        zio.Config.int.nested("max", "initial", "line", "length").withDefault(Config.default.maxInitialLineLength) ++
```

Comment:
in relation to your latest comment @42njaber, I think this kind of nesting is not what we want. Why? Because we create a lot of name spaces. In my opinion, namespaces should be few and preferable not contain only one member. Also they should logically group things.
While I can see that for `binding`, I can't see that for `max`.
My suggestion would be, that one config object in the code should relate to one namespace

Reply:
I understand and semantically that makes a lot of sense but this does not reflect the implementation of `Config` and `ConfigProvider`. I am not knowledgeable of the project to understand the intent behind the implementation, and thus what can be changed, but I can explain my view of the issue.

The current implementation is very simple :
In zio's `Config.scala`, line 76
```scala
  def nested(name: => String): Config[A] =
    Config.defer(Config.Nested(name, self))
```
And line 317
```scala
  final case class Nested[A](name: String, config: Config[A])                                    extends Composite[A]
```

In `ConfigProvide.scala`, line 268
```scala
            case Nested(name) :: tail =>
              loop(name +: path, tail)
```
And line 622
```scala
  /**
   * Constructs a ConfigProvider using a map and the specified delimiter string,
   * which determines how to split the keys in the map into path segments.
   */
  def fromMap(map: Map[String, String], pathDelim: String = ".", seqDelim: String = ","): ConfigProvider =
    fromFlat(new Flat {
      val escapedSeqDelim   = java.util.regex.Pattern.quote(seqDelim)
      val escapedPathDelim  = java.util.regex.Pattern.quote(pathDelim)
      val mapWithIndexSplit = splitIndexInKeys(map, unmakePathString, makePathString)
```

So, `.nested()` does not create any namespace, and simply defers the resolution of the config name to a config provider that has split the string along a "path" which might or might not refer to namespaces.

There is not much mechanism for case conversions, although these could be extended, but I'm getting the feeling that `.nested()` does not actually do what it is semantically meant to do, since it only matches "path segments", so it is either miss-named or is planned to be developed further in the future?

That's why I proposed different solutions, but I'm not quite sure what the intended design is, what would be useful for users, and how these features are being used in general. There seems to simply be no mechanism for distinguishing between casing and nesting in `zio`, and I'd want to have the opinion of somebody more experienced on how it *should* be, since you pointed out `.nested()` should not be used for casing conversion, only namespacing.

This bug report implies that it is intended for config providers to convert between different casings, creating compatibility between different config providers, but there seems to be no apparent generinc method for that other then using `.nested()`.

Reply:
There are other providers like Typesafe Config or AWS Parameter Store that have similar concepts to namespace. Here's an example using Typesafe Config (HOCON):

```hocon
server {
  port = 8080
  keep-alive = true
}

logging {
  root-level = INFO
  mappings {
    "some.zio.module" = WARN
  }
}
```

In this case, the server section can be described in ZIO config like `(int("port") ++ boolean("keep-alive")).nested("server")`, which has different semantic from `(int("port") ++ boolean("alive").nested("keep")).nested("server")`.